pkgs/development/compilers/opensmalltalk-vm/build-vm.nix

```diff
@@ -0,0 +1,126 @@
+, openssl
+, pango
+, pkg-config
+, xorg
+,
+}:
+let
+  owner = "OpenSmalltalk";
+  repo = "opensmalltalk-vm";
+  version = "202206021410";
+  commitHash = "c9fd36530d2c4030649e33020b7b91272a465e31";
+  srcHash= "sha256-QqElPiJuqD5svFjWrLz1zL0Tf+pHxQ2fPvkVRn2lyBI=";
+
+  src = fetchFromGitHub {
+    inherit owner repo;
+    rev = version;
+    hash = srcHash;
+  };
+in
+{ platformDir, vmName, configureFlagsArray, configureFlags }:
+stdenv.mkDerivation {
+  pname =
+    let vmNameNoDots = builtins.replaceStrings [ "." ] [ "-" ] vmName;
+    in "opensmalltalk-vm-${platformDir}-${vmNameNoDots}";
+  inherit version;
+
+  inherit src;
+
+  postPatch =
+    let
+      inherit (builtins) substring;
+      year = substring 0 4 version;
+      month = substring 4 2 version;
+      day = substring 6 2 version;
+      hour = substring 8 2 version;
+      minute = substring 10 2 version;
+      date = "${year}-${month}-${day}T${hour}:${minute}+0000";
+      abbrevHash = substring 0 12 commitHash;
+    in
+    ''
+      vmVersionDate=$(date -u '+%a %b %-d %T %Y %z' -d "${date}")
+      vmVersionFiles=$(sed -n 's/^versionfiles="\(.*\)"/\1/p' ./scripts/updateSCCSVersions)
+      for vmVersionFile in $vmVersionFiles; do
+        substituteInPlace "$vmVersionFile" \
+          --replace "\$Date\$" "\$Date: ''${vmVersionDate} \$" \
+          --replace "\$URL\$" "\$URL: ${src.url} \$" \
+          --replace "\$Rev\$" "\$Rev: ${version} \$" \
+          --replace "\$CommitHash\$" "\$CommitHash: ${abbrevHash} \$"
+      done
+      patchShebangs --build ./building/${platformDir} scripts
+      substituteInPlace ./platforms/unix/config/mkmf \
+        --replace "/bin/rm" "rm"
+      substituteInPlace ./platforms/unix/config/configure \
+        --replace "/usr/bin/file" "file" \
+        --replace "/usr/bin/pkg-config" "pkg-config" \
+    '';
+
+  preConfigure = ''
+    cd building/${platformDir}/${vmName}/build
+    ../../../../scripts/checkSCCSversion && exit 1
+    cp ../plugins.int ../plugins.ext .
+    configureFlagsArray=${configureFlagsArray}
+  '';
+
+  configureScript = "../../../../platforms/unix/config/configure";
+
+  inherit configureFlags;
+
+  buildFlags = "all";
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [
+    file
+    pkg-config
+  ];
+
+  buildInputs = [
+    alsa-lib
+    freetype
+    libpulseaudio
+    libtool
+    libuuid
+    openssl
+    pango
+    xorg.libX11
+    xorg.libXrandr
+  ];
+
+  postInstall = ''
+    rm "$out/squeak"
+    cd "$out/bin"
+    BIN="$(find ../lib -type f -name squeak)"
+    for f in $(find . -type f); do
+      rm "$f"
+      ln -s "$BIN" "$f"
+    done
+  '';
+
+  meta =
+    let
+      platform =
+        if platformDir == "linux64ARMv8" then "aarch64-linux"
+        else if platformDir == "linux64x64" then "x86_64-linux"
+        else throw "unexpected platformDir in meta.platform";
+    in
+    {
+      description = "The cross-platform virtual machine for Squeak, Pharo, Cuis, and Newspeak.";
```

Comment:
It would be ideal to set `mainProgram` here so that something like `nix run nixpkgs#opensmalltalk-vm-squeak` would just work.